cpp/src/knn/knn_opg_common.cu

```diff
@@ -532,38 +642,38 @@ void opg_knn(raft::handle_t &handle, std::vector<Matrix::Data<T> *> *out,
                           handle.get_device_allocator(), cur_batch_size, k,
                           cur_query_ptr, rowMajorIndex, rowMajorQuery);
 
-        // Synchronize before running labels copy
-        CUDA_CHECK(cudaStreamSynchronize(stream));
-
         copy_outputs(res.data(), res_I.data(), y, (size_t)cur_batch_size,
-                     (int)k, (int)n_outputs, (int)idx_desc.N, my_rank,
-                     idx_desc.partsToRanks, handle.get_device_allocator(),
-                     stream);
+                     (int)k, (int)n_outputs, (int)idx_desc.N, my_rank, idx_desc,
+                     handle.get_device_allocator(), stream);
 
         // Synchronize before sending
         CUDA_CHECK(cudaStreamSynchronize(stream));
+        CUDA_CHECK(cudaPeekAtLastError());
       }
 
-      /**
-             * Ranks exchange results.
-             * Partition owner receives. All other ranks send.
-             */
-      if (verbose)
-        std::cout << "Rank " << my_rank << ": Exchanging results" << std::endl;
-      exchange_results(res, res_I, res_D, comm, part_rank, idxRanks, stream,
-                       cur_batch_size, k, n_outputs, local_parts_completed);
+      if (part_rank == my_rank || my_rank_is_idx) {
```

Comment:
Originally, `exchange_results` was only executed when `my_rank_is_idx` was true. This was causing some crashes because for some reason (probably Dask related in the Pytests) some workers had query partition(s) without any index partition. I added `part_rank == my_rank`, so that query partition(s) only workers could receive the data and perform the reduce operation. I think that this check protects against the rare case in which a worker is partitionless.

Reply:
> for some reason (probably Dask related in the Pytests) some workers had query partition(s) without any index partition.

This is not an edge case, it's a necessary feature. It should be expected that in the worst case, the set of workers with index partitions and the set of workers with query partitions are completely disjoint. At some point (doesn't need to be in this PR), maybe we should create a pytest that verifies this. In the meantime, can you note that in a little comment in the code? It helps for other developers (including myself) to know why it's there so we don't remove it mistakingly thinking it's not being used.